File: base.database/src/main/java/alvin/base/database/dbflow/domain/repositories/PersonRepository.java
package alvin.base.database.dbflow.domain.repositories;

import com.raizlabs.android.dbflow.sql.language.Where;
import com.raizlabs.android.dbflow.sql.queriable.AsyncQuery;

import javax.inject.Inject;
import javax.inject.Singleton;

import alvin.base.database.common.domain.models.Gender;
import alvin.base.database.dbflow.domain.models.Person;
import alvin.base.database.dbflow.domain.models.Person_Table;
import alvin.lib.common.dbflow.repositories.BaseRepository;

/**
 * @see BaseRepository for more details
 */
@Singleton
public class PersonRepository extends BaseRepository<Person> {

    @Inject
    public PersonRepository() {
    }

    public AsyncQuery<Person> findById(int id) {
        return where(Person_Table.id.eq(id)).async();
    }

    public AsyncQuery<Person> findByGender(Gender gender) {
        Where<Person> where = where();
        if (gender != null) {
            where = where.and(Person_Table.gender.eq(gender));
        }
        return where.async();
    }

//    public void delete(@NonNull int id) {
//      database.executeTransaction(db -> SQLite.delete().from(Person.class).where(Person_Table.id.eq(person.getId())));
//    }
}

File: base.kotlin/src/main/java/alvin/base/kotlin/dagger/domain/repositories/PersonRepository.kt
@file:Suppress("unused")

package alvin.base.kotlin.dagger.domain.repositories

import alvin.base.kotlin.common.domain.models.Gender
import alvin.base.kotlin.common.domain.models.Person
import alvin.base.kotlin.common.domain.models.Person_Table
import alvin.lib.common.dbflow.repositories.BaseRepository
import com.raizlabs.android.dbflow.sql.language.SQLite
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class PersonRepository
@Inject constructor() : BaseRepository<Person>() {

    fun findById(id: Int, callback: (Person?) -> Unit) {
        SQLite.select().from(Person::class.java)
                .where(Person_Table.id.eq(id))
                .async()
                .querySingleResultCallback { _, person -> callback(person) }
                .execute()
    }

    fun findByGender(gender: Gender?, callback: (List<Person>) -> Unit) {
        var query = SQLite.select().from(Person::class.java).where()
        if (gender != null) {
            query = query.and(Person_Table.gender.eq(gender))
        }
        query.async()
                .queryListResultCallback { _, persons -> callback(persons) }
                .execute()
    }
}
